### PR TITLE
simple c10 implementation for std::call_once

### PR DESCRIPTION
**Pull request by jjsjann123**
A long standing bug on std::call_once: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66146
It could hang during re-entry after an exception handling.

Added a c10 implementation yielding a bulky mutex. Not the most efficient thing but at least it shouldn't hang.

### CODE REVIEW

File: aten/src/ATen/test/basic.cpp [hunk 1]
```diff
@@ -471,3 +474,48 @@ TEST(BasicTest, FactoryMethodsTest) {
     ASSERT_FALSE(tensor0.is_pinned());
   }
 }
+
+TEST(BasicTest, BasicStdTestCPU) {
+  c10::once_flag flag1, flag2;
```
Comment (jjsjann123) [inline aten/src/ATen/test/basic.cpp:479]: Note: if we switch this from c10::once_flag to std::once_flag (along witht he call_once down below), this test hangs on my machine, where i have `/usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.28`

**Comment by jjsjann123** [comment]
Not sure who to ping for review, but cc'ing @davidberard98

**Review by albanD — COMMENTED** [review]
Thanks for the fix! Only small comments.

File: c10/util/CallOnce.h [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+#include <atomic>
+#include <mutex>
+#include <thread>
+#include <utility>
+
+#include <c10/macros/Macros.h>
+
+namespace c10 {
+
+// custom c10 call_once implementation to avoid the deadlock in std::call_once.
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply c10/util/CallOnce.h:12]: I don't think it's only for ppc64, at least it also repros on my x86_64. Not sure if there's another bug that better tracks this breakage.

I'll link the bug tracker here with a note on the affected arch.

**Review by albanD — COMMENTED** [review]
cc @suo who would be the right person to provide help to add a lint rule for this?

**Comment by jjsjann123** [comment]
> cc @suo who would be the right person to provide help to add a lint rule for this?

CI looks green now. wondering if @suo has any instruction on custom lint rules to error on using `std::call_once`?

**Comment by jjsjann123** [comment]
Bumping PR.

**Comment by suo** [comment]
Would the linter just error if grep found `std::call_once`? 

It's pretty easy to define a new pattern-based linter, here is an example: https://github.com/pytorch/pytorch/blob/master/.lintrunner.toml#L348-L371

**Comment by suo** [comment]
@albanD beat me to it (*and* had a better answer lol)

**Comment by jjsjann123** [comment]
lintrunner rule added and verified that it worked locally. Thanks a lot to @albanD for the code~

Ready for review and could use a stamp :bow:

**Comment by albanD** [comment]
Thanks!
Interesting that it fails here.
Also there is an exclude field in for the rule, you might want to exclude `CallOnce.h` from that rule.

**Review by albanD — APPROVED** [review]
LGTM once the lint is fixed!

**Review by albanD — APPROVED** [review]
Thanks

**Pull request closed without merging**